src/coreclr/vm/runtimehandles.cpp

```diff
@@ -1471,362 +1453,290 @@ FCIMPL3(INT32, SignatureNative::GetParameterOffsetInternal, PCCOR_SIGNATURE sig,
+    gc.pSig->m_sig = pCorSig;
+    gc.pSig->m_cSig = cCorSig;
+    gc.pSig->m_pMethod = pMethodDesc;
 
-        if (pMethod)
-        {
-            pMethod->GetSig(&pCorSig, &cCorSig);
-            if (pMethod->GetClassification() == mcInstantiated)
-            {
-                LoaderAllocator *pLoaderAllocator = pMethod->GetLoaderAllocator();
-                if (pLoaderAllocator->IsCollectible())
-                    gc.pSig->SetKeepAlive(pLoaderAllocator->GetExposedObject());
-            }
-        }
-        else if (pFieldDesc)
-            pFieldDesc->GetSig(&pCorSig, &cCorSig);
-
-        gc.pSig->m_sig = pCorSig;
-        gc.pSig->m_cSig = cCorSig;
-        gc.pSig->m_pMethod = pMethod;
-
-        REFLECTCLASSBASEREF refDeclType = (REFLECTCLASSBASEREF)declType.GetManagedClassObject();
-        gc.pSig->SetDeclaringType(refDeclType);
-
-        PREFIX_ASSUME(pCorSig!= NULL);
-        BYTE callConv = *(BYTE*)pCorSig;
-        SigTypeContext typeContext;
-        if (pMethod)
-            SigTypeContext::InitTypeContext(
-                pMethod, declType.GetClassOrArrayInstantiation(), pMethod->LoadMethodInstantiation(), &typeContext);
-        else
-            SigTypeContext::InitTypeContext(declType, &typeContext);
+    uint32_t callConv;
+    if (FAILED(CorSigUncompressCallingConv(pCorSig, cCorSig, &callConv)))
+        COMPlusThrow(kBadImageFormatException);
 
-        MetaSig msig(pCorSig, cCorSig, pModule, &typeContext,
-            (callConv & IMAGE_CEE_CS_CALLCONV_MASK) == IMAGE_CEE_CS_CALLCONV_FIELD ? MetaSig::sigField : MetaSig::sigMember);
+    SigTypeContext typeContext;
+    if (pMethodDesc != NULL)
+    {
+        SigTypeContext::InitTypeContext(
+            pMethodDesc, declType.GetClassOrArrayInstantiation(), pMethodDesc->LoadMethodInstantiation(), &typeContext);
+    }
+    else
+    {
+        SigTypeContext::InitTypeContext(declType, &typeContext);
+    }
 
-        if (callConv == IMAGE_CEE_CS_CALLCONV_FIELD)
-        {
-            msig.NextArgNormalized();
+    Module* pModule = declType.GetModule();
+    MetaSig msig(pCorSig, cCorSig, pModule, &typeContext,
+        (callConv & IMAGE_CEE_CS_CALLCONV_MASK) == IMAGE_CEE_CS_CALLCONV_FIELD ? MetaSig::sigField : MetaSig::sigMember);
 
-            OBJECTREF refRetType = msig.GetLastTypeHandleThrowing().GetManagedClassObject();
-            gc.pSig->SetReturnType(refRetType);
-        }
-        else
-        {
-            gc.pSig->SetCallingConvention(msig.GetCallingConventionInfo());
+    if (callConv == IMAGE_CEE_CS_CALLCONV_FIELD)
+    {
+        msig.NextArgNormalized();
 
-            OBJECTREF refRetType = msig.GetRetTypeHandleThrowing().GetManagedClassObject();
-            gc.pSig->SetReturnType(refRetType);
+        OBJECTREF refRetType = msig.GetLastTypeHandleThrowing().GetManagedClassObject();
+        gc.pSig->SetReturnType(refRetType);
+    }
+    else if (pMethodDesc == NULL || !pMethodDesc->IsDynamicMethod())
```

Comment:
```suggestion
    else
```
I think `IsDynamicMethod` condition is hard to follow. We can skip more code with easier to understand condition - see my suggestion above.